File: low-level-design/design-patterns/python/behavioural/observer-pattern/uptime_service.py
from observer import Observer,UptimeAlert

class UptimeService:
    def __init__(self):
        self._observers = []

    def register_observer(self, observer: Observer):
        self._observers.append(observer)

    def unregister_observer(self, observer: Observer):
        self._observers.remove(observer)

    def notify(self, alert: UptimeAlert):
        for observer in self._observers:
            observer.update(alert)

File: low-level-design/design-patterns/golang/behavioural/observer-pattern/observer/uptime_service.go
package observer

type UptimeService struct {
	observers []Observer
}

func NewUptimeService() *UptimeService {
	return &UptimeService{}
}

func (s *UptimeService) RegisterObserver(observer Observer) {
	s.observers = append(s.observers, observer)
}

func (s *UptimeService) UnregisterObserver(observer Observer) {
	index := -1
	for i, o := range s.observers {
		if o == observer {
			index = i
			break
		}
	}
	if index != -1 {
		s.observers = append(s.observers[:index], s.observers[index+1:]...)
	}
}
func (s *UptimeService) Notify(alert UptimeAlert) {
	for _, observer := range s.observers {
		observer.Update(alert)
	}
}
